relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -1,170 +1,219 @@
+
+        SamplingConfig {
+            rules: vec![],
+            rules_v2: event_rules.chain(parent_rules).collect(),
+            // We want to take field priority on the fields from the sampling config of the project
+            // to which the incoming transaction belongs.
+            //
+            // This code ignore the situation in which we have conflicting sampling config modes
+            // between root and non-root projects.
+            mode: self.sampling_config.mode,
+        }
+    }
 }
 
+/// Checks whether unsupported rules result in a direct keep of the event or depending on the
+/// type of Relay an ignore of unsupported rules.
 fn check_unsupported_rules(
     processing_enabled: bool,
     sampling_config: &SamplingConfig,
-) -> Result<(), SamplingResult> {
-    // when we have unsupported rules disable sampling for non processing relays
+) -> Option<()> {
+    // When we have unsupported rules disable sampling for non processing relays.
     if sampling_config.has_unsupported_rules() {
         if !processing_enabled {
-            return Err(SamplingResult::Keep);
+            return None;
         } else {
             relay_log::error!("found unsupported rules even as processing relay");
         }
     }
 
-    Ok(())
-}
-
-#[derive(Clone, Debug, PartialEq)]
-struct SamplingSpec {
-    sample_rate: f64,
-    rule_id: RuleId,
-    seed: Uuid,
+    Some(())
 }
 
-fn get_trace_sampling_rule(
+/// Gets the sampling match result by creating the merged configuration and matching it against
+/// the sampling configuration.
+fn get_sampling_match_result(
     processing_enabled: bool,
-    sampling_project_state: Option<&ProjectState>,
+    project_state: &ProjectState,
+    root_project_state: Option<&ProjectState>,
     dsc: Option<&DynamicSamplingContext>,
+    event: Option<&Event>,
     ip_addr: Option<IpAddr>,
     now: DateTime<Utc>,
-) -> Result<Option<SamplingSpec>, SamplingResult> {
-    let dsc = or_ok_none!(dsc);
-
-    if sampling_project_state.is_none() {
-        relay_log::trace!("found sampling context, but no corresponding project state");
-    }
-    let sampling_project_state = or_ok_none!(sampling_project_state);
-    let sampling_config = or_ok_none!(&sampling_project_state.config.dynamic_sampling);
-    check_unsupported_rules(processing_enabled, sampling_config)?;
-
-    let rule = or_ok_none!(sampling_config.get_matching_trace_rule(dsc, ip_addr, now));
+) -> SamplingMatchResult {
+    // We get all the required data for transaction-based dynamic sampling.
+    let event = no_match_if_none!(event);
+    let sampling_config = no_match_if_none!(&project_state.config.dynamic_sampling);
+
+    // We obtain the merged sampling configuration with a concatenation of transaction rules
+    // of the current project and trace rules of the root project.
+    let merged_config = SamplingConfigs::new(sampling_config)
+        .add_root_config(root_project_state)
+        .get_merged_config();
```

Comment:
We should not clone configs / create new vectors for every event here. Ideally, we operate on a `get_merged_rules()` function that returns an iterator (not a vector) of rules.